import "library/cpp/actors/protos/interconnect.proto";
import "ydb/core/protos/base.proto";
import "ydb/core/protos/tablet.proto";
import "ydb/core/protos/subdomains.proto";
import "ydb/core/protos/flat_scheme_op.proto";
import "ydb/core/protos/follower_group.proto";

package NKikimrLocal;
option java_package = "ru.yandex.kikimr.proto";

message TTabletAvailability {
    optional NKikimrTabletBase.TTabletTypes.EType Type = 1;
    optional uint64 MaxCount = 2 [default = 1000000];
    optional int32 Priority = 3;
}

message TEvRegisterNode {
    optional fixed64 HiveId = 1;
    repeated NKikimrSubDomains.TDomainKey ServicedDomains = 2;
    optional NActorsInterconnect.TNodeLocation SystemLocation = 3;
    repeated TTabletAvailability TabletAvailability = 4;
}

message TEvRegisterNodeResult {
    optional NKikimrProto.EReplyStatus Status = 1;
    optional fixed64 HiveId = 2;
    optional uint32 HiveGeneration = 3;
}

message TLocalConfig {
    repeated NKikimrSchemeOp.TResourceProfile ResourceProfiles = 1;
}

message TEvPing {
    optional fixed64 HiveId = 1;
    optional uint32 HiveGeneration = 2;
    optional bool Purge = 3;
    optional TLocalConfig Config = 4;
}

message TEvDeadTabletAck {
    optional fixed64 TabletId = 1;
    optional uint32 Generation = 2;
    optional uint32 FollowerId = 3;
}

message TEvStatus {
    optional uint32 Status = 1;
    optional uint64 InbootTablets = 2;
    optional uint64 OnlineTablets = 3;
    optional uint64 DeadTablets = 4;
    optional uint64 AvailableWeight = 5;
    optional NKikimrTabletBase.TMetrics ResourceMaximum = 8;
    optional uint64 StartTime = 7;
}

enum EBootMode {
    BOOT_MODE_LEADER = 0;
    BOOT_MODE_FOLLOWER = 1;
}

message TEvBootTablet {
    optional NKikimrTabletBase.TTabletStorageInfo Info = 1;
    optional uint32 SuggestedGeneration = 2;
    optional EBootMode BootMode = 3 [default = BOOT_MODE_LEADER];
    optional NKikimrHive.TFollowerGroup FollowerGroup = 4;
    optional uint32 FollowerId = 5;
}

message TEvStopTablet {
    optional fixed64 TabletId = 1;
    optional uint32 FollowerId = 2;
    optional uint32 Generation = 3;
}

message TEvTabletStatus {
    optional uint32 Status = 1;
    optional fixed64 TabletID = 2;
    optional uint32 Generation = 3;
    optional uint32 FollowerId = 4;
    optional uint32 Reason = 5 [default = 34]; // ReasonError
}

message TEvEnumerateTablets {
    optional NKikimrTabletBase.TTabletTypes.EType TabletType = 1;
}

message TEvEnumerateTabletsResult {
    optional NKikimrProto.EReplyStatus Status = 1;

    message TTabletInfo {
        optional uint64 TabletId = 1;
        optional NKikimrTabletBase.TTabletTypes.EType TabletType = 2;
        optional EBootMode BootMode = 3 [default = BOOT_MODE_LEADER];
        optional uint32 FollowerId = 4;
    }

    repeated TTabletInfo TabletInfo = 2;
}

message TEvSyncTablets {
    message TTabletInfo {
        optional uint64 TabletId = 1;
        optional uint32 Generation = 2;
        optional EBootMode BootMode = 3;
        optional uint32 FollowerId = 4;
    }

    repeated TTabletInfo InbootTablets = 1;
    repeated TTabletInfo OnlineTablets = 2;
}

message TEvTabletMetricsAck {
    repeated fixed64 TabletId = 1;
    repeated uint32 FollowerId = 2;
}

message TEvReconnect {
    optional fixed64 HiveId = 1;
    optional uint32 HiveGeneration = 2;
}
